xarray/namedarray/core.py

```diff
@@ -0,0 +1,432 @@
+from __future__ import annotations
+
+import copy
+import math
+import typing
+from collections.abc import Hashable, Iterable, Mapping
+
+import numpy as np
+
+# TODO: get rid of this after migrating this class to array API
+from xarray.core import dtypes
+from xarray.core.indexing import ExplicitlyIndexed
+from xarray.core.utils import Default, _default
+from xarray.namedarray.utils import (
+    Frozen,
+    is_duck_array,
+    is_duck_dask_array,
+    to_0d_object_array,
+)
+
+# temporary placeholder for indicating an array api compliant type.
+# hopefully in the future we can narrow this down more
+T_DuckArray = typing.TypeVar("T_DuckArray", bound=typing.Any)
+
+T_NamedArray = typing.TypeVar("T_NamedArray", bound="NamedArray")
+
+
+# TODO: Add tests!
+def as_compatible_data(data, fastpath: bool = False):
+    if fastpath and getattr(data, "ndim", 0) > 0:
+        # can't use fastpath (yet) for scalars
+        return data
+
+    # TODO : check scalar
```

Comment:
please check  the scalar case.

Reply:
i'm going to work on this tomorrow morning.